### PR TITLE
Fix tests and ZipEntry DateTime Kind

### LINKED ISSUE #481: Test case failed.
### Steps to reproduce
1. CI in windows, three test cases failed:
https://ci.appveyor.com/project/icsharpcode/sharpziplib/branch/master/job/6kbl3jgxelrn1ndv

```
Errors, Failures and Warnings
1) Failed : ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(0)
  No entry matching source file "Users/appveyor/AppData/Local/Temp/1/tmpDF1A.tmp" found in archive, found "C:\Users\appveyor\AppData\Local\Temp\1\tmpDF1A.tmp"
  Expected: not null
  But was:  null
   at ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(Int32 contentLength) in C:\projects\sharpziplib\test\ICSharpCode.SharpZipLib.Tests\Zip\StreamHandling.cs:line 265
2) Failed : ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(1)
  No entry matching source file "Users/appveyor/AppData/Local/Temp/1/tmpDF7A.tmp" found in archive, found "C:\Users\appveyor\AppData\Local\Temp\1\tmpDF7A.tmp"
  Expected: not null
  But was:  null
   at ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(Int32 contentLength) in C:\projects\sharpziplib\test\ICSharpCode.SharpZipLib.Tests\Zip\StreamHandling.cs:line 265
3) Failed : ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(256)
  No entry matching source file "Users/appveyor/AppData/Local/Temp/1/tmpDF7C.tmp" found in archive, found "C:\Users\appveyor\AppData\Local\Temp\1\tmpDF7C.tmp"
  Expected: not null
  But was:  null
   at ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(Int32 contentLength) in C:\projects\sharpziplib\test\ICSharpCode.SharpZipLib.Tests\Zip\StreamHandling.cs:line 265
Run Settings
    Number of Test Workers: 2
    Work Directory: C:\projects\sharpziplib
    Internal Trace: Off
Test Run Summary
  Overall result: Failed
  Test Count: 235, Passed: 217, Failed: 3, Warnings: 0, Inconclusive: 0, Skipped: 15
    Failed Tests - Failures: 3, Errors: 0, Invalid: 0
    Skipped Tests - Ignored: 3, Explicit: 12, Other: 0
  Start time: 2020-07-04 16:30:00Z
    End time: 2020-07-04 16:30:13Z
    Duration: 12.339 seconds
```

2. Test in linux, ten test cases failed:
    The command was: `dotnet run -c Release -f netcoreapp2 -p test/ICSharpCode.SharpZipLib.TestBootstrapper/ICSharpCode.SharpZipLib.TestBootstrapper.csproj `
[test_linux_fail.log](https://github.com/icsharpcode/SharpZipLib/files/4910376/test_linux_fail.log)
```
Errors, Failures and Warnings

1) Failed : ICSharpCode.SharpZipLib.Tests.Zip.FastZipHandling.SetDirectoryModifiedDate
  Expected: 2001-01-02 00:00:00
  But was:  2001-01-02 08:00:00
   at ICSharpCode.SharpZipLib.Tests.Zip.FastZipHandling.SetDirectoryModifiedDate() in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/FastZipHandling.cs:line 612

2) Failed : ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(0)
  No entry matching source file "tmp/tmpJD38Z1.tmp" found in archive, found "/tmp/tmpJD38Z1.tmp"
  Expected: not null
  But was:  null
   at ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(Int32 contentLength) in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/StreamHandling.cs:line 267

3) Failed : ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(1)
  No entry matching source file "tmp/tmplGyoGA.tmp" found in archive, found "/tmp/tmplGyoGA.tmp"
  Expected: not null
  But was:  null
   at ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(Int32 contentLength) in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/StreamHandling.cs:line 267

4) Failed : ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(256)
  No entry matching source file "tmp/tmpL4aRm9.tmp" found in archive, found "/tmp/tmpL4aRm9.tmp"
  Expected: not null
  But was:  null
   at ICSharpCode.SharpZipLib.Tests.Zip.StreamHandling.WriteZipStreamWithNoCompression(Int32 contentLength) in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/StreamHandling.cs:line 267

5) Failed : ICSharpCode.SharpZipLib.Tests.Zip.WindowsNameTransformHandling.BasicDirectories
  Should be equal
  String lengths are both 14. Strings differ at index 8.
  Expected: "absolute\\file2"
  But was:  "absolute/file2"
  -------------------^
   at ICSharpCode.SharpZipLib.Tests.Zip.WindowsNameTransformHandling.BasicDirectories() in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs:line 123

6) Failed : ICSharpCode.SharpZipLib.Tests.Zip.WindowsNameTransformHandling.BasicFiles
  Should be equal
  Expected string length 16 but was 18. Strings differ at index 6.
  Expected: "zebidi/and/dylan"
  But was:  "zebidi\\/and\\/dylan"
  -----------------^

7) Failed : ICSharpCode.SharpZipLib.Tests.Zip.WindowsNameTransformHandling.Replacement
  Should be equal
  Expected string length 3 but was 4. Strings differ at index 1.
  Expected: "c/_"
  But was:  "c\\/_"
  ------------^

8) Failed : ICSharpCode.SharpZipLib.Tests.Zip.WindowsNameTransformHandling.ReplacementChecking
Expected an exception
   at ICSharpCode.SharpZipLib.Tests.Zip.WindowsNameTransformHandling.ReplacementChecking() in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs:line 107
...
...
...
17) Failed : ICSharpCode.SharpZipLib.Tests.Zip.ZipEntryFactoryHandling.CreatedValues
  Create time failure
  Expected: 2100-02-27 11:07:56
  But was:  2020-07-13 10:38:22.2627132
   at ICSharpCode.SharpZipLib.Tests.Zip.ZipEntryFactoryHandling.CreatedValues() in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/ZipEntryFactoryHandling.cs:line 135

18) Failed : ICSharpCode.SharpZipLib.Tests.Zip.ZipNameTransformHandling.FilenameCleaning
  Expected: 0
  But was:  21
   at ICSharpCode.SharpZipLib.Tests.Zip.ZipNameTransformHandling.FilenameCleaning() in /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master/test/ICSharpCode.SharpZipLib.Tests/Zip/ZipNameTransformHandling.cs:line 85

Run Settings
    Number of Test Workers: 4
    Work Directory: /root/workspace/hanqiang/sharpZipLib/SharpZipLib-master
    Internal Trace: Off

Test Run Summary
  Overall result: Failed
  Test Count: 235, Passed: 202, Failed: 10, Warnings: 8, Inconclusive: 0, Skipped: 15
    Failed Tests - Failures: 10, Errors: 0, Invalid: 0
    Skipped Tests - Ignored: 3, Explicit: 12, Other: 0
  Start time: 2020-07-13 02:38:13Z
    End time: 2020-07-13 02:38:23Z
    Duration: 9.158 seconds

```

### Expected behavior
Test cases all passed.

### Actual behavior
Windows three test cases failed.
linux ten test cases failed.

### Version of SharpZipLib
master branch.

### PR DESCRIPTION
**Pull request by piksel**
This should allow all tests to pass on both *nix and windows.

The current outstanding bug with ZipEntry file names not being automatically cleaned has been moved to it's own test, as this is a known (although unwanted) behaviour of the current state the code base is in.

By doing this we should be able to have failing tests prevent PRs from being merged in a more controlled manner.

Fixes #481 

_I certify that I own, and have sufficient rights to contribute, all source code and related material intended to be compiled or integrated with the source code for the SharpZipLib open source product (the "Contribution"). My Contribution is licensed under the MIT License._

### CODE REVIEW

File: test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs [hunk 1]
```diff
@@ -2,12 +2,20 @@
...
+using System.Runtime.InteropServices;
 
 namespace ICSharpCode.SharpZipLib.Tests.Zip
 {
 	[TestFixture]
 	public class WindowsNameTransformHandling : TransformBase
 	{
+		[OneTimeSetUp]
+		public void TestInit() {
+			if (Path.DirectorySeparatorChar != '\\') {
```
Comment (Numpsy) [inline test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs:14]: Is it worth using the [Platform] attribute so that the tests only run on Windows?

File: src/ICSharpCode.SharpZipLib/Zip/ZipEntry.cs [hunk 2]
```diff
@@ -742,7 +742,7 @@ public long DosTime
 					uint mon = Math.Max(1, Math.Min(12, ((uint)(value >> 21) & 0xf)));
 					uint year = ((dosTime >> 25) & 0x7f) + 1980;
 					int day = Math.Max(1, Math.Min(DateTime.DaysInMonth((int)year, (int)mon), (int)((value >> 16) & 0x1f)));
-					DateTime = new DateTime((int)year, (int)mon, day, (int)hrs, (int)min, (int)sec, DateTimeKind.Utc);
+					DateTime = new DateTime((int)year, (int)mon, day, (int)hrs, (int)min, (int)sec, DateTimeKind.Unspecified);
```
Comment (Numpsy) [inline src/ICSharpCode.SharpZipLib/Zip/ZipEntry.cs:745]: I don't see any mention of timezones in the appnote, so if it's unknown then unspecified seems a reasonable choice.

Comment (piksel) [reply src/ICSharpCode.SharpZipLib/Zip/ZipEntry.cs:745]: And more importantly, if you create a DateTime instance and specify that it is in UTC, then the OS will translate it to local time when using it to set the file attributes. If you use `Unspecified` it will not do any such translation.

File: test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs (hunk 1 above)
Comment (piksel) [reply test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs:14]: Maybe, I didn't actually know about that one. I just figured feature detection was actually what was important anyway, instead of guessing based on the OS.

Comment (Numpsy) [reply test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs:14]: I don't know what's best, just wondered if tests that involve drive letters might have issues as well as things with separators.

Comment (piksel) [reply test/ICSharpCode.SharpZipLib.Tests/Zip/WindowsNameTransformHandling.cs:14]: Yeah maybe, it's kind of hard to test until any OS starts using drive letters and forward slashes though 😁

**Comment by bastianeicher** [comment]
**TL;DR:** The `DateTimeKind.Unspecified` change looks good to me.

## Adding files to archives

When adding files to ZIP archives using `ZipFile.Add()` or `FastZip` the following code handles recording timestamps:

https://github.com/icsharpcode/SharpZipLib/blob/b5898dc3067c5503593b04afe82f8fea34924839/src/ICSharpCode.SharpZipLib/Zip/ZipEntryFactory.cs#L221-L253

The default value for the `TimeSetting` enum is `LastWriteTime`:

https://github.com/icsharpcode/SharpZipLib/blob/b5898dc3067c5503593b04afe82f8fea34924839/src/ICSharpCode.SharpZipLib/Zip/ZipEntryFactory.cs#L17-L58

This means by default the `FileInfo.CreationTime` property is read. This property provides a timestamp in local time (`DateTimeKind.Local`).

If `TimeSetting` is set to `LastWriteTimeUtc` the `FileInfo.CreationTimeUtc` property is read, which provides a timestamp in UTC time (`DateTimeKind.Utc`).

## Extracting files from archives

When extracting files from ZIP archives using `FastZip` the following code handles restoring timestamps:

https://github.com/icsharpcode/SharpZipLib/blob/b5898dc3067c5503593b04afe82f8fea34924839/src/ICSharpCode.SharpZipLib/Zip/FastZip.cs#L640-L643

The following table shows how `File.SetLastWriteTime()` (used by SharpZipLib) and `File.SetLastWriteTimeUtc()` (not used by SharpZipLib) interpret `DateTime`s with various `DateTimeKind`s:

| | `File.SetLastWriteTime()` |  `File.SetLastWriteTimeUtc()` |
|---|---|---|
| `DateTimeKind.Unspecified` | Local | UTC |
| `DateTimeKind.Local` | Local | Local |
| `DateTimeKind.Utc` | UTC | UTC |

## Impact of proposed change

This PR changes the following code to use `DateTimeKind.Unspecified` instead of `DateTimeKind.Utc` when reading timestamps from DOS time:

https://github.com/icsharpcode/SharpZipLib/blob/b5898dc3067c5503593b04afe82f8fea34924839/src/ICSharpCode.SharpZipLib/Zip/ZipEntry.cs#L745

Because the default behavior of the library is to capture timestamps as local time and `DateTimeKind.Unspecified` gets treated a s local time when writing files, this change allows timestamps to be properly round-tripped without shifting. This sounds like a good solution to me.

An alternative would be to use `DateTimeKind.Local` instead to make this more explicit. I would advise against that though, since when reading a ZIP archive its impossible to know whether its timestamps were creating from local or UTC time, so "Unspecified" seems more honest. It also allows consumers of SharpZipLib that don't use `FastZip` to choose which behavior they want by using either `File.SetLastWriteTime()` or `File.SetLastWriteTimeUtc()`.

**Comment by piksel** [comment]
Great summary!
